### ISSUE TITLE
Gaston doesn't work on Windows; gnuplot is not available on this system

### ISSUE THREAD

**Issue by xiaodaigh** [mbaz/Gaston.jl#135]
This is the error I get when I try using Gaston

```
ERROR: InitError: Gaston cannot be loaded: gnuplot is not available on this system.
Stacktrace:
 [1] error(::String) at .\error.jl:33
 [2] __init__() at C:\Users\RTX2080\.julia\packages\Gaston\JtE39\src\Gaston.jl:64
 [3] _include_from_serialized(::String, ::Array{Any,1}) at .\loading.jl:697
 [4] _require_from_serialized(::String) at .\loading.jl:748
 [5] _require(::Base.PkgId) at .\loading.jl:1039
 [6] require(::Base.PkgId) at .\loading.jl:927
 [7] require(::Module, ::Symbol) at .\loading.jl:922
 [8] top-level scope at util.jl:175
 [9] eval(::Module, ::Any) at .\boot.jl:331
 [10] eval_user_input(::Any, ::REPL.REPLBackend) at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.4\REPL\src\REPL.jl:86
 [11] run_backend(::REPL.REPLBackend) at C:\Users\RTX2080\.julia\packages\Revise\XFtoQ\src\Revise.jl:1162
 [12] top-level scope at none:0
during initialization of module Gaston
caused by [exception 1]
IOError: could not spawn `gnuplot --version`: no such file or directory (ENOENT)
Stacktrace:
 [1] _spawn_primitive(::String, ::Cmd, ::Array{Any,1}) at .\process.jl:99
 [2] #550 at .\process.jl:112 [inlined]
 [3] setup_stdios(::Base.var"#550#551"{Cmd}, ::Array{Any,1}) at .\process.jl:196
 [4] _spawn at .\process.jl:111 [inlined]
 [5] _spawn at .\process.jl:106 [inlined]
 [6] success(::Cmd) at .\process.jl:496
 [7] __init__() at C:\Users\RTX2080\.julia\packages\Gaston\JtE39\src\Gaston.jl:62
 [8] _include_from_serialized(::String, ::Array{Any,1}) at .\loading.jl:697
 [9] _require_from_serialized(::String) at .\loading.jl:748
 [10] _require(::Base.PkgId) at .\loading.jl:1039
 [11] require(::Base.PkgId) at .\loading.jl:927
 [12] require(::Module, ::Symbol) at .\loading.jl:922
 [13] top-level scope at util.jl:175
 [14] eval(::Module, ::Any) at .\boot.jl:331
 [15] eval_user_input(::Any, ::REPL.REPLBackend) at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.4\REPL\src\REPL.jl:86
 [16] run_backend(::REPL.REPLBackend) at C:\Users\RTX2080\.julia\packages\Revise\XFtoQ\src\Revise.jl:1162
 [17] top-level scope at none:0
```

**Comment by mbaz** [comment]
I don't consider this a bug in Gaston. The documentation clearly states that gnuplot is required. When gnuplot is properly installed, Gason works fine in Windows.

I will leave this bug open (with edited title) as a reminder that I need to study the artifacts scaffolding and see if it makes sense for Gaston. It's very low priority for me, though. I will accept a pull request implementing this.

**Comment by bahasalien** [comment]
get the same issued when precompiling gaston. but without poping up alarm stacktrace. after "using Gaston" repeated its works fine.

what is Operating System are you used?

**Comment by clad26** [comment]
@mbaz or anyone who's willing to give a try
```
using BinaryBuilder
BinaryBuilder.run_wizard()
for url: https://excellmedia.dl.sourceforge.net/project/gnuplot/gnuplot/5.4.0/gnuplot-5.4.0.tar.gz
After step 2a and 2b, in the terminal
cd gnuplot-5.4.0
./configure --prefix=/workspace/destdir
make 
make install
exit
```
The binary was created, although there were a some dependencies unfulfilled
But almost all of them have their binaries already created.
The dependencies I encountered for linux were 
cairo
cairo-pdf
pango
pangocairo
glib-2.0
Qt5Core
Qt5Gui
Qt5Network
Qt5Svg
Qt5PrintSupport
I think that qml.jl already has the qt binary
https://github.com/JuliaPackaging/Yggdrasil/blob/2c223a4511c715985d0ae9e6bc4cd53debfa1bae/Q/Qt/build_tarballs.jl we can skip out the other packages... only 5 are needed as dependencies for gnuplot
Also, I think that the other binaries are available too, I'm sure I saw them being installed with some other packages.
If someone can find them, just use them in step 2b

If someone has enough data, and is willing to give it a go, I hope the above instructions come in handy

**Comment by mbaz** [comment]
Does this procedure build binaries for Windows and Mac too?

**Comment by clad26** [comment]
Yes

> Does this procedure build binaries for Windows and Mac too?

Yes
Just after my linux build was over, Windows build started

**Comment by clad26** [comment]
Although I had to cancel that one as I've low on internet data here...

**Comment by mbaz** [comment]
Thanks for the pointers! I will give it a shot soon.

**Comment by mbaz** [comment]
I'm not going to build for Mac... I'm not willing to agree to any restrictions and legal terms.

**Comment by clad26** [comment]
I wanted to implement Gaston in a few Pluto tutorials... although I can't distribute it without artifacts
Gaston is just so good for this, as its lightweight and heavily equipped yet REALLY fast!

**Comment by clad26** [comment]
> I'm not going to build for Mac... I'm not willing to agree to any restrictions and legal terms.

No problem...
I just need the linux and windows build

**Comment by clad26** [comment]
> I'm not going to build for Mac... I'm not willing to agree to any restrictions and legal terms.


Actually why not just put something like
if(Sys,isapple)
 run(` /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install.sh)" `)
 run(` brew install gnuplot `)
else if(Sys.iswindows)
download(https://excellmedia.dl.sourceforge.net/project/gnuplot/gnuplot/5.2.8/gp528-win64-mingw.exe)
run(` start gp528-win64-mingw.exe `)


Since the installation process for Windows and Mac is not going to change
But for Linux/FreeBSD, the process is a bit different, the binaries would be better

**Comment by mbaz** [comment]
@clad26 That's an interesting thought! Questions:

* For apple, do you need to run those commands as root?
* Is 5.4.0 not available for Windows?
* Would it be acceptable to just print the instructions, and ask the user to execute them?

**Comment by clad26** [comment]
5.4.0 for windows was in their testing repo, release candidate 1 is out, it seems
I'd wait though, for a stable release

It seems that to install brew, we need sudo
But for installing with brew we don't

Automatically installing for the user is a better option here. We can print what we're doing and for a yes/no. If no, give an error and don't install Gaston.

**Comment by clad26** [comment]
@mbaz Thanks for taking interest！
I hope we get this feature in Gaston.jl soon！

**Comment by clad26** [comment]
@mbaz Thanks for taking interest！
I hope we get this feature in Gaston.jl soon！

**Comment by mbaz** [comment]
@clad26 Thanks for your help! I'll try to get something working soon.

**Comment by clad26** [comment]
> @clad26 Thanks for your help! I'll try to get something working soon.

Much appreciated！
https://github.com/simeonschaub/ArtifactUtils.jl
This might help

**Comment by clad26** [comment]
Also, if you're following the Windows method, if the user has an outdated version, it might be better to remove the old version！
run(｀"C:\Program Files\gnuplot\unins000.exe" ｀)
should do the trick！



Also, isn't it possible to just use the Windows build AS THE ARTIFACT？
Its much better than building it yourself every time！

**Comment by clad26** [comment]
Artifact using BinaryBuilder is only required for Linux/FreeBSD

For MAC/OSX, you might wanna look into https://github.com/JuliaPackaging/Homebrew.jl
It just outright simplifies the process of building the artifact for Mac/OSX

**Comment by clad26** [comment]
This way we can have the artifacts for the three of them....

**Comment by clad26** [comment]
Also, taking an idea from the windows build, here's the libraries we might for the linux one...
![Screenshot from 2020-10-11 17-54-16](https://user-images.githubusercontent.com/51269425/95678540-f4aa8200-0bea-11eb-8abc-734a4a072064.png)
bin folder of gnuplot windows build

**Comment by clad26** [comment]
From the windows build......
![Screenshot from 2020-10-11 17-54-16](https://user-images.githubusercontent.com/51269425/95679635-cd0ae800-0bf1-11eb-8731-3c66f52e4ff0.png)
bin of gnuplot windows
Might require the same for

**Comment by clad26** [comment]
@mbaz I think you can find the equivalent of all the above packages on https://github.com/JuliaPackaging/Yggdrasil/
Just search the name in the repo and you'll probably find it

**Comment by clad26** [comment]
I hope the above information helps

**Comment by mbaz** [comment]
@clad26 This is very useful indeed, thanks a lot! It looks like a non-trivial amount of work, though. I think I'll start with the bare minimum dependencies (i.e. no WxWidgets): I'm thinking Qt, X11, GIF, PNG, SVG, Canvas, Cairo and maybe a couple of others. I can then add other dependencies, if needed and as time allows.

I am not comfortable with modifying a user's installation, though. My plan is to have Gaston use the Julia artifact and ignore any pre-installed programs.

For Windows, just including the official binary as an artifact looks like the best solution. Still, I'm sure I have seen 5.4.0 Windows binaries, because I know I tested it before the latest Gaston release. I'll take a look and see if I can find it.

For Mac, it looks like the homebrew.jl package is the way to go. I'll give this lower priority, though, and leave it for a future release.

**Comment by clad26** [comment]
> I am not comfortable with modifying a user's installation, though. My plan is to have Gaston use the Julia artifact and ignore any pre-installed programs.

I'm sure you've heard of CUDA.jl. 
When the users install CUDA.jl, it downloads the latest version of the CUDA artifact and uses it.
But the users have an option if they want to skip the artifact download, and switch to the user's installation
Its achieved by making a folder "config" in .julia folder, and making a startup.jl file in it, and typing in startup.jl,
``` ENV["JULIA_CUDA_USE_BINARYBUILDER"]=false ```
Of course, if the user doesn't download the artifacts and doesn't have it beforehand either, it gives an error.

How about a similar mechanism for Gaston too? Ofcourse, most Julia beginners wouldn't even think about it, while more advanced Julia users will have more freedom

**Comment by clad26** [comment]
> This is very useful indeed, thanks a lot! It looks like a non-trivial amount of work, though. I think I'll start with the bare minimum dependencies (i.e. no WxWidgets): I'm thinking Qt, X11, GIF, PNG, SVG, Canvas, Cairo and maybe a couple of others. I can then add other dependencies, if needed and as time allows.
I don't think my Fedora build provides me with wxt either, so that isn't a problem personally for me. qt and x11 are sufficient for anyone.

**Comment by mbaz** [comment]
> How about a similar mechanism for Gaston too?

That'd be nice, but I don't want to try to do too much. I don't have a lot of time to devote to this right now. I prefer to get something out that works, and fine-tune over the next releases.

**Comment by clad26** [comment]
It's worth looking into
Its really easy, it seems
https://juliapackaging.github.io/BinaryBuilder.jl/dev/jll/#Non-dev'ed-JLL-packages
In your case, it would simply be 

```
UUID, from your gnuplot artifact
Gnuplot = "gnuplot"
```
right?

Well, anyway, its your call!
This would be a nice option for linux users exclusively.
For Windows users, having them use their build is pointless as it would be more or less, the same thing, or Gaston's artifact would would be more updated.

**Comment by clad26** [comment]
Also, the size of the collective artifacts matters.... It'd be great if you can keep it somewhere around 20-60mb!

**Comment by clad26** [comment]
@mbaz Thanks for taking the time for it! Heavily appreciated!
Within an year, I'll plan on starting contributing as much as I can to make it better!
Gnuplot is kinda complex, so its hard to introduce beginners to it! 
Thats really the spot where Gaston kicks in!

**Comment by clad26** [comment]
@mbaz Any chance that, if you're making the binary for linux, then can you build gnuplot 5.5？
It has a new smooth parameter which works on surfaces too. Until this, there was no easy way to do this.
If not, can you give a way to the user to access their version of gnuplot somehow？Thanks

**Comment by clad26** [comment]
@mbaz Any update on this? Were you able to get atleast the qt terminal compiled with BinaryBuilder?

**Comment by clad26** [comment]
@mbaz Any plans to make the artifact system a milestone for the 1.1 release?🙂

**Comment by mbaz** [comment]
@clad26  I've had little time to work on this, but I hope to, yes.

**Comment by clad26** [comment]
@mbaz I heard Plots was getting a Gaston backend, which I think makes the artifacts all the more important!
Any update on the articacts?

**Comment by mbaz** [comment]
It is a lot of work, and I'm unlikely to do it in the short to medium term. Another reason I'm not enthusiastic about it is that it is trivially easy to install Gnuplot in all supported platforms. I will consider a pull request if someone does the work, though.

**Comment by clad26** [comment]
@mbaz Whats your experience with wxt terminal?
I'm kinda getting close to the binary, but for now, without qt. 
I'm building wxWidgets too, for wxt terminal though 
I think that should be a good temporary alternative

**Comment by mbaz** [comment]
Wxt is a solid terminal in my experience.

**Comment by clad26** [comment]
I tried the wxt terminal on Windows yesterday
And it was smoother than qt terminal  for 5.2.8
I still don't know how it performs on Linux+GTK3 and Mac+Cocoa, but I'm expecting a similar performance...
I'm ditching the qt terminal for the wxt terminal for now...

**Comment by mbaz** [comment]
Question: let's say a user has their own installation of Gnuplot, supporting the Windows terminal, Qt, etc. Then they install Gaston with your artifact, which pulls your gnuplot binaries. Which version of Gnuplot will Gaston use in this case?

**Comment by clad26** [comment]
It'll use the artifacts I gues
Similar to CUDA.jl, I want the users to be able to choose https://juliagpu.gitlab.io/CUDA.jl/installation/overview/

```
You might want to disallow use of artifacts, e.g., because an optimized CUDA installation is available for your system. You can do so by setting the environment variable JULIA_CUDA_USE_BINARYBUILDER to false when importing CUDA.jl.

To troubleshoot discovery of a local CUDA installation, you can set JULIA_DEBUG=CUDA and see the various paths where CUDA.jl looks. By setting any of the CUDA_HOME, CUDA_ROOT or CUDA_PATH environment variables, you can guide the package to a specific directory
```

I don't wanna install CUDA twice...Its really big

**Comment by clad26** [comment]
I'll properly look into how to do this, once the binaries are built
I initially aimed to do this before 1.6, but I guess it'll take more time than I anticipated...

**Comment by clad26** [comment]
By the way, may I ask, are you a MAC user？

**Comment by mbaz** [comment]
I can go with that scheme, but ideally, it would be opt-in rather than opt-out. Gnuplot is so easy to install in all major platforms, that I think using the artifact will be the less-common scenario.

I don't use Macs at all; I use Linux.

**Comment by clad26** [comment]
If your distribution provides it, its great...
But mostly outdated... I use Fedora Linux... and am stuck with 5.2.8-6... 
5.4.1 has some cool features like 3D histograms, or so I hear
Even on Windows, they give you the official binary of 5.2.8, but you can get 5.4.1 from chocolatey...

I think the main problem here for me was kinda personal. I wrote a set of tutorials using Gaston for bar plots...
The issue here is that, I kinda gave the instructions to install on MOST platforms, in README.md！But most don't read README.md
I fear that they'll clone the Pluto notebooks, install the necessary packages, and then when the plots don't show up, they quit the browser and don't ever open it again…

**Comment by clad26** [comment]
Also, if in case their own build is somewhat messed up, thats another reason

By the way, I've the book Gnuplot in Action, but I really never read it from top to bottom. I only open it, when I need to learn something very specific, and its not in Docs, Examples, Stackoverflow, etc.

**Comment by mbaz** [comment]
I agree that some users might benefit from having the binary artifacts, and that's why I'll add this feature if it has the necessary flexibility. Thanks for working on it!

**Comment by VarLad** [comment]
@mbaz Some problems here and there.... Its almost there... I'd say anytime this month
Qt is not available on musl. It has its svg terminal and png/gif terminals. This includes distributions like Alpine Linux. Note that the binary of gnuplot provided with Alpine don't have the qt or wx terminal either

Problems on Windows 32bit. The latest version of gnuplot 5.4.1 for only 64-bit Windows. Even Chocolatey does that.
For Windows 32bit, https://sourceforge.net/projects/gnuplot/files/gnuplot/5.2.5/ can be a temporary compromise

**Comment by mbaz** [comment]
@VarLad In case you're still interested: Gaston version 2.0 is much better than 1.x, but still has no gnuplot artifact. I'd still consider a pull request if you can figure out how to build gnuplot 6.0.x, at least on Linux.

**Comment by t-bltg** [comment]
Yeah, we should target gnuplot 6.x for the artifact now.
I may be able to take a quick look at this.

**Comment by mbaz** [comment]
Building the artifact might be easier if we aim for supporting a reduced number of terminals. I think we should aim at least for these:

* qt
* pdf
* cairolatex
* svg
* sixel
* gif
* webp
* jpg
* png
* block

Am I missing any? Also, what about Windows and Mac? I think the preferred terminals in those platforms are "windows" and "aqua" respectively, but "qt" should work, at least in theory.

**Comment by t-bltg** [comment]
Those are the default terminals enable on `ubuntu`:
```
            block  Pseudo-graphics using Unicode block or Braille characters
       cairolatex  LaTeX picture environment using graphicx package and Cairo backend
           canvas  HTML Canvas object
              cgm  Computer Graphics Metafile
          context  ConTeXt with MetaFun (for PDF documents)
          domterm  DomTerm terminal emulator with embedded SVG
           dpu414  Seiko DPU-414 thermal printer [small medium large]
             dumb  ascii art for anything that prints text
              dxf  dxf-file for AutoCad (default size 120x80)
              emf  Enhanced Metafile format
         epscairo  eps terminal based on cairo
         epslatex  LaTeX picture environment using graphicx package
     epson_180dpi  Epson LQ-style 180-dot per inch (24 pin) printers
      epson_60dpi  Epson-style 60-dot per inch printers
      epson_lx800  Epson LX-800, Star NL-10, NX-1000, PROPRINTER ...
              fig  FIG graphics language V3.2 for XFIG graphics editor
              gif  GIF images using libgd and TrueType fonts
           hp500c  HP DeskJet 500c, [75 100 150 300] [rle tiff]
             hpdj  HP DeskJet 500, [75 100 150 300]
             hpgl  HP7475 and relatives [number of pens] [eject]
           hpljii  HP Laserjet series II, [75 100 150 300]
             hppj  HP PaintJet and HP3630 [FNT5X9 FNT9X17 FNT13X25]
             jpeg  JPEG images using libgd and TrueType fonts
       kittycairo  kitty in-terminal graphics using cairo
          kittygd  kitty in-terminal graphics using libgd
              lua  Lua generic terminal driver
          nec_cp6  NEC printer CP6, Epson LQ-800 [monochrome color draft]
          okidata  OKIDATA 320/321 Standard
              pbm  Portable bitmap [small medium large] [monochrome gray color]
             pcl5  PCL5e/PCL5c printers using HP-GL/2
         pdfcairo  pdf terminal based on cairo
           pict2e  LaTeX2e picture environment
              png  PNG images using libgd and TrueType fonts
         pngcairo  png terminal based on cairo
       postscript  PostScript graphics, including EPSF embedded files (*.eps)
          pslatex  LaTeX picture environment with PostScript \specials
            pstex  plain TeX with PostScript \specials
         pstricks  LaTeX picture environment with PSTricks macros
               qt  Qt cross-platform interactive terminal
          sixelgd  sixel using libgd and TrueType fonts
         sixeltek  Sixel output using bitmap graphics
            starc  Star Color Printer
              svg  W3C Scalable Vector Graphics
      tandy_60dpi  Tandy DMP-130 series 60-dot per inch graphics
          tek40xx  Tektronix 4010 and others; most TEK emulators
          tek410x  Tektronix 4106, 4107, 4109 and 420X terminals
          texdraw  LaTeX texdraw environment
             tikz  TeX TikZ graphics macros via the lua script driver
         tkcanvas  Tk canvas widget
          unknown  Unknown terminal type - not a plotting device
            vttek  VT-like tek40xx terminal emulator
             webp  single frame or animation using cairo, pango, and libwebp
              wxt  wxWidgets cross-platform interactive terminal
              x11  X11 Window System interactive terminal
             xlib  X11 Window System (dump of gnuplot_x11 command stream)
            xterm  Xterm Tektronix 4014 Mode
```
